### PR TITLE
Use BigQuery storage read API when reading external reading BigLake tables

### PR DESCRIPTION
**Pull request by anoopj**
## Description

BigQuery storage APIs support reading BigLake external tables (ie external tables with a connection). But the current implementation uses views which can be expensive, because it requires Trino issuing a SQL query against BigQuery. This PR adds support to read BigLake tables directly using the storage API.

There are no behavior changes for external tables and BQ native tables - they use the view and storage APIs respectively. Added a new test for BigLake tables.

## Additional context and related issues

Fixes https://github.com/trinodb/trino/issues/21016
https://cloud.google.com/bigquery/docs/biglake-intro

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# BigQuery
* Improve performance when reading external BigLake tables. ({issue}`21016`)
```

### CODE REVIEW

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java [hunk 1]
```diff
@@ -16,6 +16,7 @@
 import com.google.cloud.bigquery.BigQueryException;
 import com.google.cloud.bigquery.DatasetId;
 import com.google.cloud.bigquery.DatasetInfo;
+import com.google.cloud.bigquery.ExternalTableDefinition;
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:19]: Thank you for your contribution. We need a test in BaseBigQueryConnectorTest class.

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java [hunk 2]
```diff
@@ -258,12 +259,20 @@ public ConnectorTableHandle getTableHandle(ConnectorSession session, SchemaTable
             columns.add(PARTITION_DATE.getColumnHandle());
             columns.add(PARTITION_TIME.getColumnHandle());
         }
+        Optional<String> connectionId = Optional.empty();
+        if (tableInfo.get().getDefinition() instanceof ExternalTableDefinition) {
+            ExternalTableDefinition edt = tableInfo.get().getDefinition();
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:264]: Avoid abbreviations. https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#avoid-abbreviations

Also, you can use enhanced instanceof to define this variable after ExternalTableDefinition.

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java [hunk 3]
```diff
@@ -112,7 +112,10 @@ public ConnectorSplitSource getSplits(
         TableId remoteTableId = bigQueryTableHandle.asPlainTable().getRemoteTableName().toTableId();
         List<BigQuerySplit> splits = emptyProjectionIsRequired(bigQueryTableHandle.getProjectedColumns()) ?
                 createEmptyProjection(session, remoteTableId, actualParallelism, filter) :
-                readFromBigQuery(session, TableDefinition.Type.valueOf(bigQueryTableHandle.asPlainTable().getType()), remoteTableId, bigQueryTableHandle.getProjectedColumns(), actualParallelism, tableConstraint);
+                readFromBigQuery(session, TableDefinition.Type.valueOf(bigQueryTableHandle.asPlainTable().getType()),
+                        remoteTableId, bigQueryTableHandle.getProjectedColumns(),
+                        bigQueryTableHandle.asPlainTable().getConnectionId(),
+                        actualParallelism, tableConstraint);
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java:118]: Wrap per arguments:
```java
                readFromBigQuery(
                        session,
                        TableDefinition.Type.valueOf(bigQueryTableHandle.asPlainTable().getType()),
                        remoteTableId,
                        bigQueryTableHandle.getProjectedColumns(),
                        bigQueryTableHandle.asPlainTable().getConnectionId(),
                        actualParallelism,
                        tableConstraint);
```

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java [hunk 4]
```diff
@@ -121,7 +124,9 @@ private static boolean emptyProjectionIsRequired(Optional<List<BigQueryColumnHan
         return projectedColumns.isPresent() && projectedColumns.get().isEmpty();
     }
 
-    private List<BigQuerySplit> readFromBigQuery(ConnectorSession session, TableDefinition.Type type, TableId remoteTableId, Optional<List<BigQueryColumnHandle>> projectedColumns, int actualParallelism, TupleDomain<ColumnHandle> tableConstraint)
+    private List<BigQuerySplit> readFromBigQuery(ConnectorSession session, TableDefinition.Type type, TableId remoteTableId,
+                                                 Optional<List<BigQueryColumnHandle>> projectedColumns, Optional<String> connectionId,
+                                                 int actualParallelism, TupleDomain<ColumnHandle> tableConstraint)
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java:129]: Wrap per arguments:
```java
    private List<BigQuerySplit> readFromBigQuery(
            ConnectorSession session,
            TableDefinition.Type type,
            TableId remoteTableId,
            Optional<List<BigQueryColumnHandle>> projectedColumns,
            Optional<String> connectionId,
            int actualParallelism,
            TupleDomain<ColumnHandle> tableConstraint)
```

File: testing/trino-server-dev/etc/config.properties [hunk 5]
```diff
@@ -52,6 +52,7 @@ plugin.bundles=\
   ../../plugin/trino-thrift/pom.xml, \
   ../../plugin/trino-tpcds/pom.xml, \
   ../../plugin/trino-google-sheets/pom.xml, \
+  ../../plugin/trino-bigquery/pom.xml, \
```
Comment (ebyhr) [inline testing/trino-server-dev/etc/config.properties:55]: No need to change this class. You can use BigQueryQueryRunner instead.

**Review by ebyhr — CHANGES_REQUESTED** [review]

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryTableHandle.java [hunk 6]
```diff
@@ -52,12 +50,6 @@ public BigQueryTableHandle(
         this.projectedColumns = requireNonNull(projectedColumns, "projectedColumns is null");
     }
 
-    @Deprecated
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryTableHandle.java:55]: Removing this constructor seems unrelated to BigLake support. Please separate a commit.

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java (hunk 2 above)
Comment (anoopj) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:264]: Done.

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java (hunk 1 above)
Comment (anoopj) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:19]: I added a test. PTAL

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java (hunk 3 above)
Comment (anoopj) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java:118]: Done

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java (hunk 4 above)
Comment (anoopj) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java:129]: Done

File: testing/trino-server-dev/etc/config.properties (hunk 5 above)
Comment (ebyhr) [reply testing/trino-server-dev/etc/config.properties:55]: It increases bootstrap time and we don't have catalog config property file in `catalog` directory. I would recommend reverting this change.

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java (hunk 1 above)
Comment (ebyhr) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:19]: The PR isn't updated.

Comment (anoopj) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:19]: Sent it too soon. Updated.

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryTableHandle.java (hunk 6 above)
Comment (anoopj) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryTableHandle.java:55]: Moved to a separate commit.

File: testing/trino-server-dev/etc/config.properties (hunk 5 above)
Comment (anoopj) [reply testing/trino-server-dev/etc/config.properties]: Removed

File: plugin/trino-bigquery/README.md [hunk 7]
```diff
@@ -22,9 +22,16 @@ You can follow the steps below to be able to run the integration tests locally.
   **BigQuery Admin** role assigned.
 * Get the base64 encoded text of the service account credentials file using `base64
   /path/to/service_account_credentials.json`.
+* Create a new BigQuery `ClOUD_RESOURCE` connection and grant the connection service account GCS permissions.
+  [Documentation](https://cloud.google.com/bigquery/docs/create-cloud-resource-connection).
+   * `bq mk --connection --location=us --project_id=$PROJECT_ID --connection_type=CLOUD_RESOURCE $CONNECTION_ID`
+   * Now we need to grant the new connection's service account the GCS permissions.
+   * To do this, run `bq show --connection $PROJECT_ID.us.$CONNECTION_ID`, which will display the service account ID.
+   * Grant the service account GCS Object Viewer permissions. e.g.: `gsutil iam ch serviceAccount:bqcx-xxxx@gcp-sa-bigquery-condel.iam.gserviceaccount.com:objectViewer gs://DESTINATION_BUCKET_NAME`
```
Comment (hashhar) [inline plugin/trino-bigquery/README.md:30]: I assume `bqcx-xxxx@gcp-sa-bigquery-condel` is actually `$CONNECTION_ID@$PROJECT_ID`? If so please change it to be more generic otherwise it's hard to figure out what values are literals vs which need to be replaced.

**Review by hashhar — COMMENTED** [review]
looks good to me

**Comment by hashhar** [comment]
/test-with-secrets sha=9d7bd1dcad92de70856b928a99e443ec3d8b4619

File: plugin/trino-bigquery/README.md [hunk 8]
```diff
@@ -22,9 +22,16 @@ You can follow the steps below to be able to run the integration tests locally.
   **BigQuery Admin** role assigned.
 * Get the base64 encoded text of the service account credentials file using `base64
   /path/to/service_account_credentials.json`.
+* Create a new BigQuery `ClOUD_RESOURCE` connection and grant the connection service account GCS permissions.
```
Comment (findinpath) [inline plugin/trino-bigquery/README.md:25]: small typo `CLOUD_RESOURCE` instead of `ClOUD_RESOURCE` (capital `L`)

Comment (anoopj) [reply plugin/trino-bigquery/README.md]: Thank you. Fixed.

File: plugin/trino-bigquery/README.md (hunk 7 above)
Comment (anoopj) [reply plugin/trino-bigquery/README.md:30]: No. It's the output of the `bq show` command, which displays the underlying service account of the connection, on which the permissions need to be granted.  I also clarified how to set the variable while running the tests.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java [hunk 9]
```diff
@@ -690,6 +705,32 @@ public void testBigQueryExternalTable()
...
+    @Test
+    public void testBigLakeExternalTable()
+    {
+        String biglakeExternalTable = "test_biglake_external" + randomNameSuffix();
+        try {
+            onBigQuery("CREATE EXTERNAL TABLE test." + biglakeExternalTable +
+                    " WITH CONNECTION `" + bigQueryConnectionId + "`" +
+                    " OPTIONS (format = 'CSV', uris = ['gs://" + gcpStorageBucket + "/tpch/tiny/region.csv'])");
+            assertQuery("SELECT table_type FROM information_schema.tables WHERE table_schema = 'test' AND table_name = '" +
+                    biglakeExternalTable + "'", "VALUES 'BASE TABLE'");
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:717]: We could use `assertThat(getQueryRunner().tableExists(getSession(), ...))` instead. 
Same for L723-724.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java [hunk 10]
```diff
@@ -59,12 +59,15 @@ public abstract class BaseBigQueryConnectorTest
...
 
+    private String bigQueryConnectionId;
+
     @BeforeAll
     public void initBigQueryExecutor()
     {
         this.bigQuerySqlExecutor = new BigQuerySqlExecutor();
         // Prerequisite: upload region.csv in resources directory to gs://{testing.gcp-storage-bucket}/tpch/tiny/region.csv
         this.gcpStorageBucket = System.getProperty("testing.gcp-storage-bucket");
+        this.bigQueryConnectionId = System.getProperty("testing.bigquery-connection-id");
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:70]: Could you add `requireNonNull` here or within `testBigLakeExternalTable`?

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryNamedRelationHandle.java [hunk 11]
```diff
@@ -78,6 +81,12 @@ public Optional<String> getComment()
         return comment;
     }
 
+    @JsonProperty
+    public Optional<String> getConnectionId()
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryNamedRelationHandle.java:85]: nit: The String value is unused actually. boolean is sufficient.

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java [hunk 12]
```diff
@@ -134,8 +148,9 @@ private List<BigQuerySplit> readFromBigQuery(ConnectorSession session, TableDefi
             // Storage API doesn't support reading wildcard tables
             return ImmutableList.of(BigQuerySplit.forViewStream(columns, filter));
         }
-        if (type == MATERIALIZED_VIEW || type == EXTERNAL) {
+        if (type == MATERIALIZED_VIEW || (type == EXTERNAL && connectionId.isEmpty())) {
             // Storage API doesn't support reading materialized views and external tables
+            // without connection. Storage API does support BigLake external tables.
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java:153]: What happens if the connection uses non-BigLake as the source? e.g. MySQL
Screenshot 2024-03-14 at 6 45 33

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java [hunk 13]
```diff
@@ -671,6 +674,18 @@ public void testBigQuerySnapshotTable()
...
+    {
+        @Language("SQL")
+        String jobCountForTableQuery = """
+                     SELECT * FROM region-us.INFORMATION_SCHEMA.JOBS WHERE EXISTS(
+                         SELECT * FROM UNNEST(referenced_tables) AS referenced_table
+                             WHERE referenced_table.table_id = '%s')
+                    """.formatted(referencedTable);
+
+        assertEventually(() -> assertThat(bigQuerySqlExecutor.executeQuery(jobCountForTableQuery).getTotalRows())
+                .isEqualTo(expectedJobCount));
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:687]: Please leave a comment why using assertEventually.

**Comment by ebyhr** [comment]
> Support for reading BigLake tables using BigQuery storage read API. 

Please remove the following dot. https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#format-git-commit-messages

Also, I would change to `Use BigQuery storage read API when reading external reading BigLake tables` because the current title looks little misleading. Reading BigLake tables has been supported via query API.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java [hunk 14]
```diff
@@ -671,6 +674,18 @@ public void testBigQuerySnapshotTable()
         }
     }
 
+    private void assertJobCountForTable(String referencedTable, long expectedJobCount)
```
[the remark this answers is not in the document]
Comment (anoopj) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java]: Done.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java (hunk 9 above)
Comment (anoopj) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java]: Done. I had to set the schema to `TEST_SCHEMA` in BigQueryRunner for this to work. Hope that's OK.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java (hunk 10 above)
Comment (anoopj) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java]: Done

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryNamedRelationHandle.java (hunk 11 above)
Comment (anoopj) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryNamedRelationHandle.java]: Fixed

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java (hunk 12 above)
Comment (anoopj) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java:153]: You won't be able to create an external table with a non-BigLake connections such as Cloud SQL.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java (hunk 13 above)
Comment (anoopj) [reply plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java]: Done

**Comment by anoopj** [comment]
> > Support for reading BigLake tables using BigQuery storage read API.
> 
> Please remove the following dot. https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#format-git-commit-messages
> 
> Also, I would change to `Use BigQuery storage read API when reading external reading BigLake tables` because the current title looks little misleading. Reading BigLake tables has been supported via query API.

Done.

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java (hunk 12 above)
Comment (ebyhr) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java:153]: Can you share the source? For instance, [Create object tables](https://cloud.google.com/bigquery/docs/object-tables) contains the example with connection. This is different page from BigLake tables.

Comment (ebyhr) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java:153]: I'm not asking if storage APIs support BigLake tables or not. You said "You won't be able to create an external table with a non-BigLake connections". However, https://cloud.google.com/bigquery/docs/object-tables indicates that connection-id can be used for other tables.

Comment (anoopj) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java:153]: For BigLake tables, the connection type has to be BigLake (a.k.a. `CLOUD_RESOURCE`): that was your original question. Yes, the BigLake connection could also be used for creating object tables. I don't think object tables are particularly interesting from a Trino perspective: they are primarily used for unstructured data analysis using BQML.

**Comment by anoopj** [comment]
@ebyhr Do you have any more feedback or can this be merged?

**Comment by anoopj** [comment]
@ebyhr @hashhar Friendly ping here. We have a GCP customer who is waiting for this PR to be merged.

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java (hunk 12 above)
Comment (ebyhr) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java:153]: >  I don't think object tables are particularly interesting from a Trino perspective

My question is the behavior when accessing object tables with this change. It's not covered by tests in my understanding.

**Comment by hashhar** [comment]
see https://github.com/trinodb/trino/pull/21017#discussion_r1536270950, I think it's an important question.

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java (hunk 12 above)
Comment (anoopj) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java:153]: Object tables are not supported by the storage read API because there is no use case for it currently. I will exclude object tables from this code path, so that it goes through the query path. This requires some [changes](https://github.com/googleapis/java-bigquery/issues/3216) in the BigQuery Java client library to expose the `objectMetadata` field and we will need to upgrade Trino to the newer client libraries. I will contribute that change and update this PR.

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BigQueryQueryRunner.java [hunk 15]
```diff
@@ -121,7 +121,7 @@ public static Session createSession()
     {
         return testSessionBuilder()
                 .setCatalog("bigquery")
-                .setSchema(TPCH_SCHEMA)
+                .setSchema(TEST_SCHEMA)
```
Comment (ssheikin) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BigQueryQueryRunner.java:124]: unrelated change?

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryNamedRelationHandle.java [hunk 16]
```diff
@@ -78,6 +81,12 @@ public Optional<String> getComment()
         return comment;
     }
 
+    @JsonProperty
+    public boolean hasConnection()
+    {
+        return connectionId.isPresent();
```
Comment (ssheikin) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryNamedRelationHandle.java:87]: There is no need to keep `Optional<String>` as a field if we are interested only in boolean.

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java [hunk 17]
```diff
@@ -134,8 +148,9 @@ private List<BigQuerySplit> readFromBigQuery(ConnectorSession session, TableDefi
             // Storage API doesn't support reading wildcard tables
             return ImmutableList.of(BigQuerySplit.forViewStream(columns, filter));
         }
-        if (type == MATERIALIZED_VIEW || type == EXTERNAL) {
+        if (type == MATERIALIZED_VIEW || (type == EXTERNAL && !hasConnection)) {
```
[the remark this answers is not in the document]
Comment (ssheikin) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java:151]: Also please consider to split this condition to two `if`s: one for `MATERIALIZED_VIEW` and the second one for the `EXTERNAL`

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java [hunk 18]
```diff
@@ -59,12 +60,17 @@ public abstract class BaseBigQueryConnectorTest
     protected BigQuerySqlExecutor bigQuerySqlExecutor;
     private String gcpStorageBucket;
```
Comment (ssheikin) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:62]: this empty line is not needed.

File: plugin/trino-bigquery/README.md [hunk 19]
```diff
@@ -22,9 +22,16 @@ You can follow the steps below to be able to run the integration tests locally.
...
+
 * The `TestBigQueryWithDifferentProjectIdConnectorSmokeTest` requires an alternate project ID which is different from the
   project ID attached to the service account but the service account still has access to.
-* Set the VM options `bigquery.credentials-key`, `testing.gcp-storage-bucket`, and `testing.alternate-bq-project-id` in the IntelliJ "Run Configuration"
-  (or on the CLI if using Maven directly). It should look something like
-  `-Dbigquery.credentials-key=base64-text -Dtesting.gcp-storage-bucket=DESTINATION_BUCKET_NAME -Dtesting.alternate-bq-project-id=bigquery-cicd-alternate`.
+  * Set the VM options `bigquery.credentials-key`, `testing.gcp-storage-bucket`, `testing.alternate-bq-project-id` and `testing.bigquery-connection-id` in the IntelliJ "Run Configuration"
+    (or on the CLI if using Maven directly). It should look something like
+        `-Dbigquery.credentials-key=base64-text -Dtesting.gcp-storage-bucket=DESTINATION_BUCKET_NAME -Dtesting.alternate-bq-project-id=bigquery-cicd-alternate -testing.bigquery-connection-id=my_project.us.connection-id`.
```
Comment (ssheikin) [inline plugin/trino-bigquery/README.md:36]: `-` -> `-D`

File: plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java [hunk 20]
```diff
@@ -59,12 +60,17 @@ public abstract class BaseBigQueryConnectorTest
...
+
     @BeforeAll
     public void initBigQueryExecutor()
     {
         this.bigQuerySqlExecutor = new BigQuerySqlExecutor();
         // Prerequisite: upload region.csv in resources directory to gs://{testing.gcp-storage-bucket}/tpch/tiny/region.csv
         this.gcpStorageBucket = System.getProperty("testing.gcp-storage-bucket");
+        this.bigQueryConnectionId = requireNonNull(
+                System.getProperty("testing.bigquery-connection-id"),
+                "testing.bigquery-connection-id is not set");
```
Comment (ssheikin) [inline plugin/trino-bigquery/src/test/java/io/trino/plugin/bigquery/BaseBigQueryConnectorTest.java:73]: Usually we have such code in trino in one line. 
Also, how is it populated on CI?
I see you've updated documentation with `testing.bigquery-connection-id=my_project.us.connection-id`, but there are no corresponding changes within `./.github/workflows/ci.yml` as they are for
`bigquery.credentials-key`, `testing.gcp-storage-bucket`, `testing.alternate-bq-project-id`

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java (hunk 12 above)
Comment (ssheikin) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java:153]: @anoopj What is your estimation on when newer BigQuery client libraries will be ready to be included to Trino?

Comment (anoopj) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java:153]: It may take a couple of weeks.

Comment (ssheikin) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java:153]: @anoopj do you have any update?

Comment (anoopj) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java:153]: My [PR](https://github.com/googleapis/java-bigquery/pull/3217) was merged last week. I'm waiting for the new release of the client libraries, which is scheduled to happen next week.

Comment (anoopj) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java:153]: Just a heads-up. I'm busy with other projects and won't be able to come back to this in the next few weeks.

Comment (ssheikin) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java:153]: @anoopj could you please share your plans regarding this PR?

**Comment by velascoluis** [comment]
Can we get this merged?

**Comment by hashhar** [comment]
@anoopj Do you plan to continue this? Or should someone else pick this up and drive to completion? I see that the newer client is released already.

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java (hunk 12 above)
Comment (anoopj) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java:153]: @ssheikin Sorry for the delay. I am not planning to work on this in the near term. Did you need this feature?

**Comment by anoopj** [comment]
@hashhar I am not planning to work on this.

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java (hunk 12 above)
Comment (ssheikin) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQuerySplitManager.java:153]: Thank you! Yeah, we were interested in it. Do you know if there is someone who is willing to take over this PR?

**Comment by mosabua** [comment]
@ssheikin and @hashhar .. are you taking this over here or in a new PR? Should we close this one?

**Comment by k-haley1** [comment]
> @ssheikin and @hashhar .. are you taking this over here or in a new PR? Should we close this one?

please leave it open for now. we are discussing.

**Comment by Praveen2112** [comment]
@mosabua I could take up the work on this PR

**Comment by mosabua** [comment]
Sounds good @Praveen2112 .. since @anoopj is not going to continue you can continue on this PR or start a new one with his work. Just link to this PR if you create a new one.

**Comment by marcinsbd** [comment]
Continuation of this PR - https://github.com/trinodb/trino/pull/22974
